### PR TITLE
Initial Governance contract integration

### LINKED ISSUE #14: dBFT stage 4: integrate voting
Header chain should be verifiable on its own, but we don't want to include the set of not-often-changed keys into every block, so:
 * allow for a key list to be included into `extra` as well, but only use it when `nextConsensus` differs from the current set
 * it should be included into the genesis block
 * once it's in the header, take them, verify they equal the `nextConsensus` hash and use to verify the next block
 * CNs should get the list from the voting contract once it's ready

### PR DESCRIPTION
**Pull request by AnnaShaleva**
Here's a draft that I have right now, several questions need to be decided, they are marked as TODOs in the code. Close #14.

~~The good news is that the code is able to call Governance API and get the current set of validators. The bad news is that sometimes contract call doesn't work as expected, the following error occurs on attempt to unpack the result, I'm investigating it.~~
UPD: the described problem is solved, now Governance calls work as expected. Now I'm working on the rest of less critical TODOs:
- [x] Decide about final contract hash (ref. https://github.com/bane-labs/go-ethereum/pull/146#discussion_r1519684969).
- [x] ~~Automate contract build and genesis allocations pre-filling.~~ Will be done in #149.
- [x] Make contract code network-independent (ref. https://github.com/bane-labs/go-ethereum/pull/146#discussion_r1519655264 and https://github.com/bane-labs/go-ethereum/pull/146#discussion_r1519662507) and update the rest of privnet/testnet configurations afterwards.
- [x] Deal with genesis-based contract call processing.
- [x] Deal with next epoch validators calculations (ref. https://github.com/bane-labs/go-ethereum/pull/146#discussion_r1519970256).
- [x] Add validators cache.

### CODE REVIEW

**Comment by AnnaShaleva** [comment]
@chenquanyu, @txhsl welcome to review, but please keep in mind that it's a very early version of integration.

@nicolegys, this PR is not ready for testing yet, I'll notify you as soon as it's ready.

**Comment by AnnaShaleva** [comment]
@roman-khimov, first version is ready for review. The contract version should be updated (currently the very old one is used to test 7-nodes privnet), but the idea of the PR won't be changed if no issues found.

File: consensus/dbft/dbft.go [hunk 1]
```diff
@@ -1589,56 +1610,81 @@ func (c *DBFT) APIs(chain consensus.ChainHeaderReader) []rpc.API {
...
+		return nil, fmt.Errorf("failed to unpack validators: %w", err)
+	}
+	var res = make([]common.Address, len(valSet))
+	for i := range valSet {
+		res[i] = valSet[i]
+	}
+
+	// TODO: roman-khimov, we've agreed that validators are sorted using lexicographic order, but in contract they are
+	// sorted by votes. Need to come to some single solution.
+	slices.SortFunc(res, common.Address.Cmp)
```
Comment (chenquanyu) [inline consensus/dbft/dbft.go:1653]: Do we need to sort validators in contract? It seems to be enough to sort them here.

File: core/systemcontracts/contracts.go [hunk 2]
```diff
@@ -2,7 +2,10 @@ package systemcontracts
...
..."address"},{"internalType":"uint256","name":"","type":"uint256"}],"name":"unclaimedEpochsOf","outputs":[{"internalType":"uint256","name":"","type":"uint256"}],"stateMutability":"view","type":"function"},{"inputs":[{"internalType":"address","name":"candidateTo","type":"address"}],"name":"vote","outputs":[],"stateMutability":"payable","type":"function"},{"inputs":[{"internalType":"address","name":"","type":"address"},{"internalType":"uint256","name":"","type":"uint256"}],"name":"votedAmount","outputs":[{"internalType":"uint256","name":"","type":"uint256"}],"stateMutability":"view","type":"function"},{"inputs":[{"internalType":"uint256","name":"","type":"uint256"}],"name":"votedCandidates","outputs":[{"internalType":"uint256","name":"","type":"uint256"}],"stateMutability":"view","type":"function"},{"inputs":[{"internalType":"address","name":"","type":"address"},{"internalType":"uint256","name":"","type":"uint256"}],"name":"votedTo","outputs":[{"internalType":"address","name":"","type":"address"}],"stateMutability":"view","type":"function"},{"inputs":[],"name":"voterWithdraw","outputs":[],"stateMutability":"nonpayable","type":"function"},{"stateMutability":"payable","type":"receive"}]`
```
Comment (chenquanyu) [inline core/systemcontracts/contracts.go]: May be we only need the interface abi here.

File: consensus/dbft/dbft.go (hunk 1 above)
Comment (AnnaShaleva) [reply consensus/dbft/dbft.go:1653]: > Do we need to sort validators in contract?

We don't. But I'd like to ensure, that it's OK for us to sort validators on dBFT engine side.

File: core/systemcontracts/contracts.go (hunk 2 above)
Comment (AnnaShaleva) [reply core/systemcontracts/contracts.go]: Yes, it will be updated once we have Governance finalized.

File: consensus/dbft/dbft.go (hunk 1 above)
Comment (AnnaShaleva) [reply consensus/dbft/dbft.go:1653]: It's OK for us, so let this sorting code to be present here. TODO is removed.

File: core/state_processor.go [hunk 3]
```diff
@@ -189,3 +196,28 @@ func ProcessBeaconBlockRoot(beaconRoot common.Hash, vmenv *vm.EVM, statedb *stat
 	_, _, _ = vmenv.Call(vm.AccountRef(msg.From), *msg.To, msg.Data, 30_000_000, common.U2560)
 	statedb.Finalise(true)
 }
+
+// ProcessOnPersist applies a system call to the governance contract.
+func ProcessOnPersist(vmenv *vm.EVM, statedb *state.StateDB) error {
```
Comment (chenquanyu) [inline core/state_processor.go:201]: Now we perform a call to OnPerist method of Governance contract via direct state modification (without fully-qualified transaction creation; for free) before in-block transactions processing. I think we can raise an issue first and we can implement free transactions in the future.

Comment (AnnaShaleva) [reply core/state_processor.go:201]: Ref. #152.

**Comment by AnnaShaleva** [comment]
@roman-khimov, ready for review & merge, @nicolegys is performing 7-nodes privnet testing. Should we wait for test results before the merge?

File: consensus/dbft/dbft.go [hunk 4]
```diff
@@ -1593,7 +1601,8 @@ func (c *DBFT) APIs(chain consensus.ChainHeaderReader) []rpc.API {
 // to be sorted by bytes order (even if returned from governance contract).
 func (c *DBFT) getNextBlockValidators(blockHash common.Hash, blockNum uint64, compute bool) ([]common.Address, error) {
 	// Currently we don't have governance contract, thus, always return standby set.
-	if true {
+	// TODO: to be removed.
```
Comment (roman-khimov) [inline consensus/dbft/dbft.go]: TODOs in the final commit series?

File: consensus/dbft/dbft.go [hunk 5]
```diff
@@ -1603,26 +1612,33 @@ func (c *DBFT) getNextBlockValidators(blockHash common.Hash, blockNum uint64, co
...
-	}
+	method := "getCurrentConsensus" // current epoch validators
+	/*if compute {
+		// TODO: given time-dependent `EPOCH_DURATION`, calculation of next consensus becomes complicated. We need
+		// `getNextConsensus` API to be presented in Governance contract since `getConsensus` is not enough.
+		// `getConsensus` calculates consensus based on the current state, even if epoch is not finished yet. What we need
+		// is `getNextConsensus` that returns current epoch validators for the middle of dBFT epoch and next epoch validators
+		// for the last block in the epoch. We can organize this at the code-level, but it complicates getNextBlockValidators call, so
+		// why not to move it to the contract level where it costs nothing.
+		// @roman-khimov, agree?
```
Comment (roman-khimov) [inline consensus/dbft/dbft.go]: TODOs in the final commit series?

File: consensus/dbft/dbft.go [hunk 6]
```diff
@@ -1603,26 +1612,33 @@ func (c *DBFT) getNextBlockValidators(blockHash common.Hash, blockNum uint64, co
...
+		method = "getNextConsensus" // current epoch validators for the middle of dBFT epoch and next epoch validators for the last block in epoch
+	}*/
 
 	ctx, cancel := context.WithCancel(context.Background())
-	// Cancel when we are finished consuming integers
+	// Cancel when we are finished consuming integers.
```
Comment (roman-khimov) [inline consensus/dbft/dbft.go]: Mixed with other changes.

File: consensus/dbft/dbft.go [hunk 7]
```diff
@@ -1632,9 +1648,19 @@ func (c *DBFT) getNextBlockValidators(blockHash common.Hash, blockNum uint64, co
 		return nil, err
 	}
 
-	var valSet []common.Address
+	var valSet [7]common.Address
```
Comment (roman-khimov) [inline consensus/dbft/dbft.go]: Can we get this number from the contract?

File: consensus/dbft/dbft.go [hunk 8]
```diff
@@ -455,9 +451,16 @@ func New(config *params.DBFTConfig, _ ethdb.Database) (*DBFT, error) {
 				log.Crit("failed to finalize and assemble proposal",
 					"err", err)
 			}
-
```
Comment (roman-khimov) [inline consensus/dbft/dbft.go:456]: Whitespace.

File: consensus/dbft/dbft.go [hunk 9]
```diff
@@ -573,6 +573,27 @@ func New(config *params.DBFTConfig, _ ethdb.Database) (*DBFT, error) {
...
 				return false
 			}
+
+			// Verify NextConsensus based on the state got after in-block transactions processing. Make a
+			// state copy in order to avoid state modifications potentially made by getValidators call.
+			// The original state will be committed if block is accepted.
+			// TODO: technically, the contract itself doesn't change the state in getCurrentValidators
+			// (at least, at the current version), and we don't need to create a copy. But I'm not sure
+			// whether processing engine makes changes in the state, it should be investigated in a separate
+			// issue. And if not, then state may not be copied.
```
Comment (roman-khimov) [inline consensus/dbft/dbft.go]: Number?

File: internal/ethapi/api.go [hunk 10]
```diff
@@ -1138,6 +1138,12 @@ func DoCall(ctx context.Context, b Backend, args TransactionArgs, blockNrOrHash
 	return doCall(ctx, b, args, state, header, overrides, blockOverrides, timeout, globalGasCap)
 }
 
+func DoCallAtState(ctx context.Context, b Backend, args TransactionArgs, state *state.StateDB, header *types.Header, overrides *StateOverride, blockOverrides *BlockOverrides, timeout time.Duration, globalGasCap uint64) (*core.ExecutionResult, error) {
```
Comment (roman-khimov) [inline internal/ethapi/api.go:1143]: Undocumented?

File: core/systemcontracts/contracts.go [hunk 11]
```diff
@@ -1,11 +1,18 @@
...
...tputs":[{"internalType":"uint256","name":"","type":"uint256"}],"stateMutability":"view","type":"function"},{"inputs":[{"internalType":"address","name":"candidateTo","type":"address"}],"name":"vote","outputs":[],"stateMutability":"payable","type":"function"},{"inputs":[{"internalType":"address","name":"","type":"address"}],"name":"voteHeight","outputs":[{"internalType":"uint256","name":"","type":"uint256"}],"stateMutability":"view","type":"function"},{"inputs":[{"internalType":"address","name":"","type":"address"}],"name":"votedAmount","outputs":[{"internalType":"uint256","name":"","type":"uint256"}],"stateMutability":"view","type":"function"},{"inputs":[{"internalType":"address","name":"","type":"address"}],"name":"votedTo","outputs":[{"internalType":"address","name":"","type":"address"}],"stateMutability":"view","type":"function"},{"inputs":[{"internalType":"address","name":"","type":"address"}],"name":"voterGasPerVote","outputs":[{"internalType":"uint256","name":"","type":"uint256"}],"stateMutability":"view","type":"function"},{"inputs":[],"name":"withdrawRegisterFee","outputs":[],"stateMutability":"nonpayable","type":"function"},{"stateMutability":"payable","type":"receive"}]`))
```
Comment (roman-khimov) [inline core/systemcontracts/contracts.go]: You can do this in `init` and panic.

**Comment by roman-khimov** [comment]
We need to test voting that changes the set of validators, btw.

**Comment by nicolegys** [comment]
CNs failed when trying to change epoch.

```
CRIT [03-27|08:16:07.046] failed to process state from proposal    err="failed to process block: could not apply OnPersist [0x1774aa018de852e473c1d5cf332fc22ebfe7af960b8d7eb64370e9fa20c4354d]: onPersist call failed: execution reverted" number=12 "seal hash"=c98122..e91b71 "parent hash"=0x40f5b34ede642c91ea99563ef4f0bbe7d384604fa4c6d2db4c39df4e2e7bc517 "intermediate merkle root"=047e7d..96635e coinbase=0x1212000000000000000000000000000000000003 "gas limit"=30,000,000 "gas used"=0 difficulty=1 "mix digest"=0x0000000000000000000000000000000000000000000000000000000000000000 nonce=3 time=1,711,527,227 "uncle hash"=0x1dcc4de8dec75d7aab85b567b6ccd41ad312451b948a7413f0a142fd40d49347 txs=0
```

**Comment by txhsl** [comment]
> CNs failed when trying to change epoch.

Have you tried to register at least 7 different addresses as candidates before epoch change? The related contract method is https://github.com/bane-labs/go-ethereum/blob/governance/contracts/solidity/Governance.sol#L119.

**Comment by AnnaShaleva** [comment]
> CNs failed when trying to change epoch.

Node expects the contract to properly handle `onPersist` call, and if it's not the case, then the node have to stop operation. The case is that it's Primary node who is trying to construct PrepareRequest, and if the construction fails, then we're toast.

> Have you tried to register at least 7 different addresses as candidates before epoch change?

So I'd expect the contract to handle the case when not enough candidates are registered. In this case StandBy validators should be returned. Because we will have such cases in the real network.

File: consensus/dbft/dbft.go (hunk 4 above)
Comment (AnnaShaleva) [reply consensus/dbft/dbft.go]: These TODOs are resolved in the subsequent commits. Fixed.

Just FYI, we have a set of similar commits in NeoGo :D

File: consensus/dbft/dbft.go (hunk 5 above)
Comment (AnnaShaleva) [reply consensus/dbft/dbft.go]: Fixed.

File: consensus/dbft/dbft.go (hunk 6 above)
Comment (AnnaShaleva) [reply consensus/dbft/dbft.go]: Moved to a separate commit.

File: consensus/dbft/dbft.go (hunk 7 above)
Comment (AnnaShaleva) [reply consensus/dbft/dbft.go]: It was a kind of an intermediate contract version, it's fixed in the later commits. But now when we have a relatively stable contract, we don't need a code for intermediate version, so removed.

File: consensus/dbft/dbft.go (hunk 8 above)
Comment (AnnaShaleva) [reply consensus/dbft/dbft.go:456]: Fixed.

File: consensus/dbft/dbft.go (hunk 9 above)
Comment (AnnaShaleva) [reply consensus/dbft/dbft.go]: [Number](https://github.com/bane-labs/go-ethereum/issues/154).

File: internal/ethapi/api.go (hunk 10 above)
Comment (AnnaShaleva) [reply internal/ethapi/api.go:1143]: I followed the existing `DoCall` style, it also doesn't have a comment. But probably we may write better code. Fixed.

File: core/systemcontracts/contracts.go (hunk 11 above)
Comment (AnnaShaleva) [reply core/systemcontracts/contracts.go]: Fixed.

**Review by roman-khimov — APPROVED** [review]

**Pull request merged**